### ISSUE TITLE
Compatibility modes

### ISSUE THREAD

**Issue by vbohata** [elastic/elasticsearch#37453]
Large number of applications do not keep up with current Elasticsearch versions. This is not only issues with major versions (5, 6, ...) but some applications are even certified to be running with some minor ES version. It would be nice to have some compatibility modes available in Elasticsearch - index based or user/role based. So even most current elasticsearch could behave to some clients like 6.2.0 or 5.4.0, ... I know this would be hard for implementation but very very useful feature in enterprise.

**Comment by vbohata** [comment]
This is more a feature request. It would be nice for example to be able to create elasticsearch-5 index in elasticsearch 6/7 with all the features like elasticsearch 5 in that time or have for example 6.2.0 behaviour even with 6.2.0 non-security bugs if their fix can cause incompatibility. The reason for this is to have most current supported ES cluster supporting older index formats/features instead of having multiple (sometimes non-supported) ES clusters.

**Comment by jasontedor** [comment]
Thanks @vbohata for your interest in Elasticsearch and for opening this for discussion. As developers ourselves, we understand the inherent conflict in dealing with breaking changes and keeping up with latest. As maintainers, we have to weigh consider the benefits of each feature against the costs, in accordance with our goals for Elasticsearch.

We want to keep Elasticsearch moving forward as fast as possible. We are committed to this, and have aggressive plans into the future. We also want users to be able to come with us, so there's the tension of moving fast, yet doing it in a way that users can keep up with the changes. The more that break, the more potential there is for users to stay behind.

We discussed your proposal during our weekly team discussion meeting today and we feel the balance is not right. It would cost us tremendously to maintain backwards compatibility guarantees of the sort that you are looking for, at the expense of moving the project forward. Instead, we strike a balance. We have committed to bringing users with us across major version boundaries by investing in rolling upgrades across major version boundaries, and being far more deliberate about breaking changes than when the project was far less mature than it is now (frankly, immature compared to where we are now). Along the lines of being more deliberate, not only are we more reluctant to make breaking changes, but when we do we offer deprecation warnings, migration assistants, and in some cases backport the new functionality to the previous major to help users transition. We think these changes help us stay committed to moving forward fast, while also making it easy for users to come with us. And for users that can not come with us (which can happen in slower moving enterprises), we have a long maintenance policy for the previous major release. Today, we are still committing bug fixes to 5.6 and doing periodic releases. This brings back critical bug fixes including for security issues for our users that need to continue to use 5.6.

Wrapping all of this up, on the spectrum of backwards compatibility forever/moving forward slowly to breaking anything all the time/moving forward as fast as possible, we are happy with our current balance of the pace at which we can move yet still bring users with us. Given our current philosophy, we do not want to move further down the spectrum towards moving more slowly and incurring the expense of offering generous backwards compatibility guarantees. The offset for that is our maintenance policy for the previous major.

I am empathetic this might not satisfy you, as it means your feature request is being closed without being fulfilled. I at least hope that you understand where we are coming from.
